browser/ui/webui/brave_rewards/rewards_panel_handler.cc

```diff
@@ -0,0 +1,70 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/webui/brave_rewards/rewards_panel_handler.h"
+
+#include <utility>
+
+#include "brave/browser/brave_rewards/rewards_service_factory.h"
+#include "brave/components/brave_rewards/browser/rewards_service.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_finder.h"
+
+RewardsPanelHandler::RewardsPanelHandler(
+    mojo::PendingRemote<brave_rewards::mojom::Panel> panel,
+    mojo::PendingReceiver<brave_rewards::mojom::PanelHandler> receiver,
+    base::WeakPtr<ui::MojoBubbleWebUIController::Embedder> embedder,
+    Profile* profile,
+    brave_rewards::RewardsPanelCoordinator* panel_coordinator)
+    : receiver_(this, std::move(receiver)),
+      panel_(std::move(panel)),
+      embedder_(embedder),
+      profile_(profile),
+      panel_coordinator_(panel_coordinator) {
+  DCHECK(embedder_);
+  DCHECK(profile_);
+  if (panel_coordinator_) {
+    panel_observation_.Observe(panel_coordinator_);
+  }
+}
+
+RewardsPanelHandler::~RewardsPanelHandler() = default;
+
+void RewardsPanelHandler::ShowUI() {
+  if (embedder_) {
+    embedder_->ShowUI();
+  }
+}
+
+void RewardsPanelHandler::CloseUI() {
+  if (embedder_) {
+    embedder_->CloseUI();
+  }
+}
+
+void RewardsPanelHandler::StartRewards(StartRewardsCallback callback) {
+  auto* rewards_service =
+      brave_rewards::RewardsServiceFactory::GetForProfile(profile_);
```

Comment:
same thing as above applies here. It's better to avoid dependencies on factories in general when possible

Reply:
Would you suggest that `RewardsPanelUI` obtain the reference to `RewardsService` and pass this to `RewardsPanelHandler`?

Reply:
Updated `RewardsPanelHandler` to accept a `RewardsService*` on construction instead of a `Profile*`.